metagraph/src/graph/annotated_graph_algorithm.cpp

```diff
@@ -15,249 +12,416 @@ namespace graph {
+                    count_merge(union_mask, sig, path);
+
+                if (found_out_round2)
+                    count_merge(union_mask, sig, path, 1);
             }
-        }
-    }, num_threads);
+        }, num_threads);
 
-    return std::make_unique<bitmap_vector>(std::move(unitig_mask));
-}
+        std::atomic_thread_fence(std::memory_order_acquire);
 
-std::unique_ptr<bitmap_vector>
-mask_nodes_by_unitig_labels(const AnnotatedDBG &anno_graph,
-                            const std::vector<Label> &labels_in,
-                            const std::vector<Label> &labels_out,
-                            size_t num_threads,
-                            double label_mask_in_fraction,
-                            double label_mask_out_fraction,
-                            double label_other_fraction) {
-    const auto &dbg = anno_graph.get_graph();
-    const auto &annotation = anno_graph.get_annotation();
-    const auto &label_encoder = annotation.get_label_encoder();
-
-    const double label_in_factor = label_mask_in_fraction * labels_in.size();
-    const double label_out_factor = label_mask_out_fraction * labels_out.size();
-
-    bool detected_missing_labels = false;
-    std::unordered_set<uint64_t> labels_in_enc;
-    for (const auto &label_in : labels_in) {
-        try {
-            labels_in_enc.emplace(label_encoder.encode(label_in));
-        } catch (...) {
-            logger->error("Mask-in label {} is not found in annotation", label_in);
-            detected_missing_labels = true;
-        }
+        masked_graph->set_mask(new bitmap_vector(std::move(union_mask)));
     }
-    std::unordered_set<uint64_t> labels_out_enc;
-    for (const auto &label_out : labels_out) {
-        try {
-            labels_out_enc.emplace(label_encoder.encode(label_out));
-        } catch (...) {
-            logger->error("Mask-out label {} is not found in annotation", label_out);
-            detected_missing_labels = true;
+
+    // Filter unitigs from masked graph based on filtration criteria
+    logger->trace("Filtering out background");
+
+    if (config.label_mask_in_unitig_fraction == 0.0
+            && config.label_mask_out_unitig_fraction == 1.0
+            && config.label_mask_other_unitig_fraction == 1.0) {
+        if (config.label_mask_in_kmer_fraction == 0.0
+                && config.label_mask_out_kmer_fraction == 1.0) {
+            logger->trace("Bypassing background filtration");
+            return masked_graph;
         }
+
+        logger->trace("Filtering by node");
+        update_masked_graph_by_node(*masked_graph, [&](node_index node) {
+            uint64_t in_count = count_vector.first[node * 2];
+            uint64_t out_count = count_vector.first[node * 2 + 1];
+            uint64_t sum = in_count + out_count;
+            return in_count >= config.label_mask_in_kmer_fraction * sum
+                && out_count <= config.label_mask_out_kmer_fraction * sum;
+        }, num_threads);
+
```

Comment:
Why do you write all these callbacks? I see no point in them here. They bring no benefits.
Why not do a pass over `count_vector` in a loop and initialize everything directly?

Reply:
So you basically don't need function `update_masked_graph_by_node`.